npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb

```diff
@@ -502,46 +631,109 @@ def git_ssh_requirements_to_swap
           @git_ssh_requirements_to_swap
         end
 
-        def post_process_npm_lockfile(original_content, updated_content)
-          updated_content =
-            replace_project_metadata(updated_content, original_content)
+        def post_process_npm_lockfile(original_content, updated_content, lockfile_name)
+          updated_content = replace_project_metadata(updated_content, original_content)
 
           # Switch SSH requirements back for git dependencies
+          updated_content = replace_swapped_git_ssh_requirements(updated_content)
+
+          # Switch from details back for git dependencies (they will have
+          # changed because we locked them)
+          updated_content = replace_locked_git_dependencies(updated_content)
+
+          # Switch back npm 7 lockfile "pacakages" requirements from the package.json
+          updated_content = restore_locked_package_dependencies(lockfile_name, updated_content)
+
+          # Switch back the protocol of tarball resolutions if they've changed
+          # (fixes an npm bug, which appears to be applied inconsistently)
+          replace_tarball_urls(updated_content)
+        end
+
+        # NOTE: This is a workaround to "sync" what's in package.json
+        # requirements and the `packages.""` entry in npm 7 v2 lockfiles. These
+        # get out of sync because we lock git dependencies (that are not being
+        # updated) to a specific sha to prevent unrelated updates and the way we
+        # invoke the `npm install` cli, where we might tell npm to install a
+        # specific versionm e.g. `npm install eslint@1.1.8` but we keep the
+        # `package.json` requirement for eslint at `^1.0.0`, in which case we
+        # need to copy this from the manifest to the lockfile after the update
+        # has finished.
+        def restore_locked_package_dependencies(lockfile_name, lockfile_content)
+          npm_version = Dependabot::NpmAndYarn::Helpers.npm_version(lockfile_content)
```

Comment:
We check for the `npm_version` 6 times in this file, I know we talked about extracting an npm 7 updater, but maybe we can start with:

```ruby
def npm7?
  return @npm7 if defined?(@npm7)

  @npm7 = Dependabot::NpmAndYarn::Helpers.npm_version(lockfile_content)
end
```

And using that everywhere? Might be a micro-optimization but I think it slightly clarifies the code and should be a pretty cheap change?

Reply:
Yeah this would be good but I think we need to refactor the lockfile updater a bit to get there as we currently pass around the lockfile name everywhere from here: https://github.com/dependabot/dependabot-core/blob/720e663fa95b952421f0883cdefe2f3048a6e797/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb#L26

Initialising the class with a single lockfile to be updated would make the class a lot simpler.

Reply:
Extracted it into a method but still passing around the content